### PR TITLE
Integrate chronos

### PR DESCRIPTION
**Pull request by markstory**
This integrates cakephp/chronos in, and removes Carbon. In addition there are few other important changes that could be considered backwards incompatible, and at the very least will need to be documented:
- Date and DateTime columns return Immutable objects. This changes how one interacts with date/datetime objects. You can no longer `$date->modify('+1 minute')` for example. Instead you _must_ do `$date = $date->modify('+1 minute');`.
- A new `Cake\I18n\Date` class has been added. This class is used for Date columns, and presents the same interface as `Cake\I18n\Time`.
## Open Questions
- Are the changes related to Immutable objects something we are comfortable with?
- Should we include shims/aliases that allow userland code to continue to hint against `Carbon\Carbon`?
- Do the Date column changes seem like a risky change?

### CODE REVIEW

File: src/I18n/Date.php [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+     * will be used for formatting the date part of the object and the second position
+     * will be used to format the time part.
+     *
+     * @var string|array|int
+     * @see \Cake\I18n\DateFormatTrait::i18nFormat()
+     */
+    protected static $_toStringFormat = [IntlDateFormatter::SHORT, -1];
+
+    /**
+     * The format to use when formatting a time using `Cake\I18n\Time::timeAgoInWords()`
```
Comment (antograssiot) [inline src/I18n/Date.php]: using ``Cake\I18n\Date::timeAgoInWords()`` instead ?

File: src/Database/Type/DateTimeType.php [hunk 2]
```diff
@@ -126,7 +127,7 @@ public function toPHP($value, Driver $driver)
      */
     public function marshal($value)
     {
-        if ($value instanceof DateTime) {
+        if ($value instanceof DateTime || $value instanceof DateTimeImmutable) {
```
Comment (lorenzo) [inline src/Database/Type/DateTimeType.php]: DatetimeInterface will work better here

**Comment by lorenzo** [comment]
My idea was to also offer a mutable `Time` implementation that people could opt into when upgrading. We can also document how to wire the `Type` classes so they retain the old behavior if necessary.

I think we need to add a shim for Carbon inside Chronos itself. That will also help attract people to the library.

Another thing I was thinking about is to make the `timeAgoInWords` implementation a separate object, that can be instantiated inside the date classes themselves. That way we can provide a migration path for user coming from Carbon that need their own implementation of `diffForHumans()`

File: tests/bootstrap.php [hunk 3]
```diff
@@ -101,6 +101,11 @@
 ]);
 
 Log::config([
+    // 'queries' => [
+    //     'className' => 'Console',
+    //     'stream' => 'php://stderr',
+    //     'scopes' => ['queriesLog']
+    // ],
```
Comment (lorenzo) [inline tests/bootstrap.php]: Was this intentional?

Comment (markstory) [inline tests/bootstrap.php]: Somewhat. I frequently enable query logging and wanted a quick way to enable it.

**Comment by markstory** [comment]
> I think we need to add a shim for Carbon inside Chronos itself. That will also help attract people to the library.

Ok, I can take care of that.

What did you have in mind for allowing mutable types? We could do something like:

```php
Type::build('datetime')->useImmutable();
Type::build('date')->useImmutable();
```

And default to mutable object creation. Defaulting to mutable objects would be more backwards compatible for existing applications. In the 3.2 compatible app skeleton we could default to immutable objects though.

**Comment by markstory** [comment]
The test failures seem to be caused by intl versions, as the same tests pass fine for me locally. I'll try changing the test expectation around.

File: tests/bootstrap.php (hunk 3 above)
Comment (lorenzo) [inline tests/bootstrap.php]: Good stuff, then :)

**Comment by lorenzo** [comment]
@markstory I think the type classes have a setter (or maybe it was a public property?) for setting the datetime class to use. We can just point users to those options. I think the default skeleton does not  need to change as the type classes are now wired to the immutable classes directly

**Comment by markstory** [comment]
While this pull request is defaulting to immutable dates, should we default to mutable ones for backwards compatibility reasons? Right now if an application is doing any mutations those changes will be lost. Any existing type checks will also fail. If we wanted to be more conservative, we could default to mutable, and enable immutability in the app-skeleton.

**Comment by lorenzo** [comment]
@markstory Being conservative would also meant that `Time` needs to extend from `MutableDateTime` instead of `Chronos`. But I think we can start having mutable dates in the Type classes for 3.2 and then change the default in 3.3?

**Comment by markstory** [comment]
I've switched most of the code to be more compatible. `Cake\I18n\Time` is now inheriting from `MutableDateTime`. There are still a few failing tests related to dates not being mutable. I need to add a `MutableDate` to chronos for better compatibility, and will be doing that tomorrow.

**Comment by jhli** [comment]
Why are mutable time objects bad? (Just asking)?

**Comment by markstory** [comment]
@jhli The main benefit is that immutable objects are much harder to make mistakes with. For example:

```php
$thing->created->modify('+1 day');
```

Won't be persisted by the ORM right now, as the property didn't 'change'. Immutable objects make accidental side-effects impossible. The above example shows just one of the many ways datetimes could be modified in-place perhaps by accident. With an immutable object modifications are always explicit:

```php
$thing->created = $thing->created->modify('+1 day');
```

**Comment by jhli** [comment]
Thanks - good point.

**Comment by lorenzo** [comment]
Great job @markstory !

**Comment by markstory** [comment]
Thanks @lorenzo :smile:

**Comment by markstory** [comment]
I can look into providing immutable objects next.

**Pull request merged**